File: js/templates/outbound-ops/lib/ops.ts
/**
 * OutboundOps — the daily-ops logic as an importable class, so you can use it
 * in your own service instead of only from the CLI.
 *
 *   import { OutboundOps } from "./lib/ops.js"
 *   const ops = new OutboundOps(client)      // your own configured client
 *   const brief = await ops.brief()          // structured data — no printing
 *   const diag  = await ops.diagnose({ domains: ["acme.com"] })  // read-only
 *   if (!diag.ready) await ops.contain(diag) // execute containment on YOUR call
 *
 * The CLI commands (commands/brief.ts, commands/incident.ts) are thin wrappers
 * over this class. Formatting lives in lib/render.ts — this file never prints.
 */
import {
  createInstantlyClient,
  getCampaignAnalytics,
  pauseCampaign,
  patchAccount,
  type InstantlyClient,
} from "@instantly-ai/sdk"
import { senderHealth, repliesWaiting, overview, itemsOf, type SenderHealth } from "./reads.js"
import { verifyDns, type DnsResult } from "./dns.js"
import { thresholds as defaultThresholds, type Thresholds } from "./thresholds.js"

export interface Action { priority: number; text: string; command?: string }
export interface BriefResult {
  health: SenderHealth
  replies: number | null
  overview: unknown
  actions: Action[]
}
export interface CampaignScore { id: string; name: string; sent: number; bounceRate: number; replyRate: number }
export type ContainmentAction =
  | { kind: "pause-campaign"; label: string; campaignId: string }
  | { kind: "drop-daily-limit"; label: string; email: string; dailyLimit: number }
export interface Diagnosis {
  health: SenderHealth
  campaigns: CampaignScore[]
  worstCampaign?: CampaignScore
  dns: DnsResult[]
  causes: string[]
  plan: ContainmentAction[] // recommended, NOT executed — inspect it, then call contain()
  ready: boolean // true = nothing to contain
}
export interface ContainmentResult { action: ContainmentAction; ok: boolean; error?: unknown }

function num(obj: unknown, keys: string[]): number | undefined {
  const o = obj as Record<string, unknown> | null
  for (const k of keys) { const v = o?.[k]; if (typeof v === "number") return v }
  return undefined
}
const emptyHealth = (): SenderHealth => ({ total: null, green: 0, flagged: [] })

export class OutboundOps {
  private client: InstantlyClient
  private thresholds: Thresholds

  constructor(client: InstantlyClient, opts: { thresholds?: Thresholds } = {}) {
    this.client = client
    this.thresholds = opts.thresholds ?? defaultThresholds
  }

  /** Convenience for scripts: build from INSTANTLY_API_KEY (+ optional AS_WORKSPACE). */
  static fromEnv(): OutboundOps {
    const apiKey = process.env.INSTANTLY_API_KEY
    if (!apiKey) throw new Error("INSTANTLY_API_KEY is not set")
    const asWorkspace = process.env.AS_WORKSPACE
    return new OutboundOps(
      createInstantlyClient({ apiKey, ...(asWorkspace ? { defaultHeaders: { "x-as-workspace": asWorkspace } } : {}) }),
    )
  }

  /** The morning picture — structured, no side effects. Reads degrade independently. */
  async brief(): Promise<BriefResult> {
    const health = await senderHealth(this.client).catch(emptyHealth)
    const replies = await repliesWaiting(this.client).catch(() => null)
    const ov = await overview(this.client).catch(() => null)
    const actions: Action[] = []
    if (health.flagged.length > 0) {
      actions.push({ priority: 1, text: `${health.flagged.length} sender(s) unhealthy — deliverability risk`, command: "ops.diagnose() → contain()  ·  npm run incident" })
    }
    if (replies && replies > 0) {
      actions.push({ priority: 2, text: `${replies} replies waiting`, command: "Instantly Unibox → app.instantly.ai" })
    }
    return { health, replies, overview: ov, actions }
  }

  /** Diagnose deliverability trouble. READ-ONLY — returns an inspectable containment plan; does not execute it. */
  async diagnose(opts: { domains?: string[] } = {}): Promise<Diagnosis> {
    const health = await senderHealth(this.client).catch(emptyHealth)
    const raw = await getCampaignAnalytics(this.client, { query: {} }).catch(() => null)
    const campaigns: CampaignScore[] = ((itemsOf(raw) ?? (Array.isArray(raw) ? raw : [])) as unknown[])
      .map((c) => {
        const co = c as Record<string, unknown>
        const id = (co.campaign_id ?? co.id ?? "") as string
        const name = (co.campaign_name ?? co.name ?? id) as string
        const sent = num(co, ["emails_sent_count", "sent", "emails_sent"]) ?? 0
        const bounced = num(co, ["bounced", "bounces", "bounced_count"]) ?? 0
        const replies = num(co, ["total_replies", "replies", "reply_count"]) ?? 0
        return { id, name, sent, bounceRate: sent > 0 ? bounced / sent : 0, replyRate: sent > 0 ? replies / sent : 0 }
      })
      .filter((c) => c.id)
    const eligible = campaigns.filter((c) => c.sent >= 50)
    const worstCampaign = [...eligible].sort((a, b) => b.bounceRate - a.bounceRate)[0]
    const dns = opts.domains?.length ? await verifyDns(opts.domains) : []

    const causes: string[] = []
    const highBounce = eligible.filter((c) => c.bounceRate >= this.thresholds.bounceCrit)
    if (highBounce.length >= 2) causes.push("List quality — high bounce across multiple campaigns; verify/refresh the list before more sends.")
    else if (worstCampaign && worstCampaign.bounceRate >= this.thresholds.bounceCrit) causes.push(`Campaign "${worstCampaign.name}" bounce ${(worstCampaign.bounceRate * 100).toFixed(1)}% — likely a bad list segment or a cold domain.`)
    if (health.flagged.length) causes.push(`${health.flagged.length} sender(s) in a bad state — deliverability/reputation.`)
    for (const d of dns) if (d.spf !== "ok" || d.dmarc !== "ok") causes.push(`Authentication — ${d.domain}: SPF ${d.spf}, DMARC ${d.dmarc} (fix DNS; DKIM needs your selector).`)
    if (!causes.length) causes.push("No clear cause from Instantly-side signals — check external reputation (Postmaster / MXToolbox).")

    const plan: ContainmentAction[] = []
    if (worstCampaign && worstCampaign.bounceRate >= this.thresholds.bounceCrit) {
      plan.push({ kind: "pause-campaign", label: `Pause "${worstCampaign.name}" (bounce ${(worstCampaign.bounceRate * 100).toFixed(1)}%)`, campaignId: worstCampaign.id })
    }
    for (const f of health.flagged) {
      plan.push({ kind: "drop-daily-limit", label: `Drop ${f.email} daily_limit → 25`, email: f.email, dailyLimit: 25 })
    }
    return { health, campaigns, worstCampaign, dns, causes, plan, ready: plan.length === 0 }
  }

  /** Execute a diagnosis's containment plan (pause worst campaign, drop limits). This WRITES. */
  async contain(diagnosis: Diagnosis): Promise<ContainmentResult[]> {
    const results: ContainmentResult[] = []
    for (const a of diagnosis.plan) {
      try {
        if (a.kind === "pause-campaign") await pauseCampaign(this.client, { path: { id: a.campaignId } })
        else await patchAccount(this.client, { path: { email: a.email }, body: { daily_limit: a.dailyLimit } })
        results.push({ action: a, ok: true })
      } catch (error) {
        results.push({ action: a, ok: false, error })
      }
    }
    return results
  }
}

File: python/templates/outbound-ops/app/ops.py
"""OutboundOps — the daily-ops logic as an importable class, so you can use it
in your own service instead of only from the CLI.

    from app.ops import OutboundOps
    ops = OutboundOps(client)                 # your own configured client
    brief = ops.brief()                        # structured data — no printing
    diag = ops.diagnose(domains=["acme.com"])  # read-only
    if not diag["ready"]:
        ops.contain(diag)                      # execute containment on YOUR call

The CLI commands (app/brief.py, app/incident.py) are thin wrappers over this.
Formatting lives in app/render.py — this module never prints.
"""
import os

from instantly import create_instantly_client, InstantlyApiError
from instantly.resources.campaign import get_campaign_analytics, pause_campaign
from instantly.resources.account import patch_account

from .reads import sender_health, replies_waiting, overview, items_of
from .dns import verify_dns
from .thresholds import THRESHOLDS


def _num(obj, keys):
    if isinstance(obj, dict):
        for k in keys:
            v = obj.get(k)
            if isinstance(v, (int, float)):
                return v
    return None


def _empty_health():
    return {"total": None, "green": 0, "flagged": []}


def _safe(fn, fallback):
    try:
        return fn()
    except Exception:
        return fallback


class OutboundOps:
    def __init__(self, client, thresholds=None):
        self.client = client
        self.thresholds = thresholds or THRESHOLDS

    @classmethod
    def from_env(cls):
        """Convenience for scripts: build from INSTANTLY_API_KEY (+ optional AS_WORKSPACE)."""
        api_key = os.environ.get("INSTANTLY_API_KEY")
        if not api_key:
            raise RuntimeError("INSTANTLY_API_KEY is not set")
        as_ws = os.environ.get("AS_WORKSPACE")
        client = (create_instantly_client(api_key, default_headers={"x-as-workspace": as_ws})
                  if as_ws else create_instantly_client(api_key))
        return cls(client)

    def brief(self):
        """The morning picture — structured, no side effects."""
        health = _safe(lambda: sender_health(self.client), _empty_health())
        replies = _safe(lambda: replies_waiting(self.client), None)
        ov = _safe(lambda: overview(self.client), None)
        actions = []
        if health["flagged"]:
            actions.append({"priority": 1, "text": f"{len(health['flagged'])} sender(s) unhealthy — deliverability risk",
                            "command": "ops.diagnose() → contain()  ·  python main.py incident"})
        if replies and replies > 0:
            actions.append({"priority": 2, "text": f"{replies} replies waiting",
                            "command": "Instantly Unibox → app.instantly.ai"})
        return {"health": health, "replies": replies, "overview": ov, "actions": actions}

    def diagnose(self, domains=None):
        """Diagnose deliverability trouble. READ-ONLY — returns an inspectable plan; does not execute it."""
        health = _safe(lambda: sender_health(self.client), _empty_health())
        raw = _safe(lambda: get_campaign_analytics(self.client, {"query": {}}), None)
        campaigns = []
        for c in (items_of(raw) or (raw if isinstance(raw, list) else [])):
            co = c if isinstance(c, dict) else {}
            cid = co.get("campaign_id") or co.get("id")
            if not cid:
                continue
            sent = _num(co, ["emails_sent_count", "sent", "emails_sent"]) or 0
            bounced = _num(co, ["bounced", "bounces", "bounced_count"]) or 0
            replies = _num(co, ["total_replies", "replies", "reply_count"]) or 0
            campaigns.append({"id": cid, "name": co.get("campaign_name") or co.get("name") or cid, "sent": sent,
                              "bounce_rate": (bounced / sent) if sent else 0.0,
                              "reply_rate": (replies / sent) if sent else 0.0})
        eligible = [c for c in campaigns if c["sent"] >= 50]
        worst = max(eligible, key=lambda c: c["bounce_rate"], default=None)
        dns = verify_dns(domains) if domains else []

        causes = []
        high_bounce = [c for c in eligible if c["bounce_rate"] >= self.thresholds["bounce_crit"]]
        if len(high_bounce) >= 2:
            causes.append("List quality — high bounce across multiple campaigns; verify/refresh the list before more sends.")
        elif worst and worst["bounce_rate"] >= self.thresholds["bounce_crit"]:
            causes.append(f'Campaign "{worst["name"]}" bounce {worst["bounce_rate"] * 100:.1f}% — likely a bad list segment or a cold domain.')
        if health["flagged"]:
            causes.append(f"{len(health['flagged'])} sender(s) in a bad state — deliverability/reputation.")
        for d in dns:
            if d.get("spf") != "ok" or d.get("dmarc") != "ok":
                causes.append(f"Authentication — {d['domain']}: SPF {d['spf']}, DMARC {d['dmarc']} (fix DNS; DKIM needs your selector).")
        if not causes:
            causes.append("No clear cause from Instantly-side signals — check external reputation (Postmaster / MXToolbox).")

        plan = []
        if worst and worst["bounce_rate"] >= self.thresholds["bounce_crit"]:
            plan.append({"kind": "pause-campaign", "label": f'Pause "{worst["name"]}" (bounce {worst["bounce_rate"] * 100:.1f}%)', "campaign_id": worst["id"]})
        for f in health["flagged"]:
            plan.append({"kind": "drop-daily-limit", "label": f"Drop {f['email']} daily_limit → 25", "email": f["email"], "daily_limit": 25})
        return {"health": health, "campaigns": campaigns, "worst_campaign": worst, "dns": dns,
                "causes": causes, "plan": plan, "ready": len(plan) == 0}

    def contain(self, diagnosis):
        """Execute a diagnosis's containment plan (pause worst campaign, drop limits). This WRITES."""
        results = []
        for a in diagnosis["plan"]:
            try:
                if a["kind"] == "pause-campaign":
                    pause_campaign(self.client, {"path": {"id": a["campaign_id"]}})
                else:
                    patch_account(self.client, {"path": {"email": a["email"]}, "body": {"daily_limit": a["daily_limit"]}})
                results.append({"action": a, "ok": True})
            except InstantlyApiError as e:
                results.append({"action": a, "ok": False, "error": f"API {e.status}"})
        return results
